### PR TITLE
[WIP] External memory pool allocator

### PR DESCRIPTION
**Pull request by hvy**
More object oriented variant of https://github.com/cupy/cupy/pull/1883

### CODE REVIEW

File: cupy/__init__.py [hunk 1]
```diff
@@ -685,13 +685,7 @@ def get_array_module(*args):
...
-# set default allocator
-_default_memory_pool = cuda.MemoryPool()
-_default_pinned_memory_pool = cuda.PinnedMemoryPool()
-
-cuda.set_allocator(_default_memory_pool.malloc)
-cuda.set_pinned_memory_allocator(_default_pinned_memory_pool.malloc)
+_default_pinned_memory_pool = None
 
 
 def get_default_memory_pool():
```
Comment (niboshi) [inline cupy/__init__.py]: I think this function does not have to be touched.

Comment (hvy) [reply cupy/__init__.py]: Should we keep `_default_memory_pool` ?

Comment (niboshi) [reply cupy/__init__.py]: Yes, it doesn't have to with this refactoring, I guess.

**Comment by hvy** [comment]
Closed in favor of https://github.com/cupy/cupy/pull/1904.

**Pull request closed without merging**